### PR TITLE
DOP-4916: Implement SoftwareSourceCode structured data

### PR DESCRIPTION
**Pull request by rayangler**
### Stories/Links:

DOP-4916

### Current Behavior:

* [Different language examples](https://www.mongodb.com/docs/atlas/driver-connection/)
* [IO code block examples](https://www.mongodb.com/docs/atlas/atlas-search/compound/)
* [No language example](https://www.mongodb.com/docs/atlas/billing/)

### Staging Links:

Google doesn't care about `SoftwareSourceCode`, so they're not present in Rich Results.

* [Different language examples](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/sandbox/cloud-docs/raymund.rodriguez/DOP-4916-software-source-code/driver-connection/index.html)
  * [Validator](https://validator.schema.org/#url=https%3A%2F%2Fdocs-mongodb-org-stg.s3.us-east-2.amazonaws.com%2Fsandbox%2Fcloud-docs%2Fraymund.rodriguez%2FDOP-4916-software-source-code%2Fdriver-connection%2Findex.html)
* [IO code block examples](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/sandbox/cloud-docs/raymund.rodriguez/DOP-4916-software-source-code/atlas-search/compound/index.html)
  * [Validator](https://validator.schema.org/#url=https%3A%2F%2Fdocs-mongodb-org-stg.s3.us-east-2.amazonaws.com%2Fsandbox%2Fcloud-docs%2Fraymund.rodriguez%2FDOP-4916-software-source-code%2Fatlas-search%2Fcompound%2Findex.html)
* [No language example](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/sandbox/cloud-docs/raymund.rodriguez/DOP-4916-software-source-code/billing/index.html)
  * [Validator](https://validator.schema.org/#url=https%3A%2F%2Fdocs-mongodb-org-stg.s3.us-east-2.amazonaws.com%2Fsandbox%2Fcloud-docs%2Fraymund.rodriguez%2FDOP-4916-software-source-code%2Fbilling%2Findex.html)

### Notes:

* Implements `SoftwareSourceCode` structured data for code blocks, with `programmingLanguage` as optional. A new mapping of different full language names has been added here to help map languages from code blocks to more SEO-friendly names. Since we don't currently have a formal list of supported languages, I'm basing this off of what I've seen and what LeafyGreen supports.
* IO code blocks have been updated so that they're always in the HTML regardless if the output is shown on the screen or not. This allows robots to see the content, and the structured data together without needing interaction.

Ideally, these would be separate PRs, but I've added a few bonus changes based on related work. Let me know if you'd like me to create separate PRs!
* **Bonus:** Removed unnecessary IDs from structured data scripts. I originally added it to `VideoObject` because the Docs Landing SD had it, but it seems like it's not actually used for anything. Less code to worry about!
* **Bonus:** Restructured new `VideoObject` to adopt new `StructuredData` class introduced in #1249.
* **Bonus:** Fixed `BreadcrumbContainer` prop type. This reduces noisy console errors for local dev.

### README updates

- - [ ] This PR introduces changes that should be reflected in the README, and I have made those updates.
- - [x] This PR does not introduce changes that should be reflected in the README

### CODE REVIEW

File: src/components/Video/index.js [hunk 1]
```diff
@@ -121,10 +111,13 @@ const Video = ({ nodeData: { argument, options = {} } }) => {
 
   return (
     <>
-      {hasAllReqFields && (
-        <script id={`video-object-sd-${url}`} type="application/ld+json">
-          {JSON.stringify(structuredData)}
-        </script>
+      {videoObjectSd.isValid() && (
```
Comment (seungpark) [inline src/components/Video/index.js:114]: Matt had suggested doing this in the memo and returning ! 👍

File: src/utils/get-language.js [hunk 2]
```diff
@@ -15,3 +53,20 @@ export const getLanguage = (lang) => {
...
+/**
+ * @param {string} lang The language passed to the code block directive
+ * @returns {string | undefined} The formal name of the language, if it exists
+ */
+export const getFullLanguageName = (lang) => {
+  if (!lang) {
+    return undefined;
+  }
+
+  const fullLangName = LANGUAGE_NAMES[lang];
```
Comment (seungpark) [inline src/utils/get-language.js:66]: [minor] - can we check for hasOwnProperty here for cases like 'constructor' 😬

File: src/components/Code/Code.js [hunk 3]
```diff
@@ -85,77 +86,89 @@ const Code = ({
     reportAnalytics('CodeblockCopied', { code });
   }, [code]);
 
+  const softwareSourceCodeSd = useMemo(() => new SoftwareSourceCodeSd({ code, lang }), [code, lang]);
```
Comment (seungpark) [inline src/components/Code/Code.js:89]: same minor point here for checking validity and returning !

**Review by seungpark — COMMENTED** [review]
LGTM - left some minor comments below

note - CMIIW but i think there were no styling changes (just shifted from structured data HTML) !

File: src/components/Video/index.js (hunk 1 above)
Comment (rayangler) [reply src/components/Video/index.js]: Makes sense. Thanks for flagging (and to Matt for suggesting)

File: src/utils/get-language.js (hunk 2 above)
Comment (rayangler) [reply src/utils/get-language.js]: I hope nobody is using `constructor` as a programming language. Done

File: src/components/Code/Output.js [hunk 4]
```diff
@@ -27,36 +28,50 @@ const OutputContainer = styled.div`
...
-        showLineNumbers={linenos}
-        darkMode={true}
-        copyable={false}
-        lineNumberStart={lineno_start}
+    <>
+      {softwareSourceCodeSd && (
+        <script
+          type="application/ld+json"
+          dangerouslySetInnerHTML={{
+            __html: softwareSourceCodeSd.toString(),
```
Comment (mmeigs) [inline src/components/Code/Output.js:46]: Do we need toString() here? Isn't it already stringified above?

File: src/components/Video/index.js [hunk 5]
```diff
@@ -121,10 +111,13 @@ const Video = ({ nodeData: { argument, options = {} } }) => {
...
     <>
-      {hasAllReqFields && (
-        <script id={`video-object-sd-${url}`} type="application/ld+json">
-          {JSON.stringify(structuredData)}
-        </script>
+      {videoObjectSd && (
+        <script
+          type="application/ld+json"
+          dangerouslySetInnerHTML={{
+            __html: videoObjectSd.toString(),
```
Comment (mmeigs) [inline src/components/Video/index.js:118]: ditto

File: src/components/Code/Output.js (hunk 4 above)
Comment (rayangler) [reply src/components/Code/Output.js]: Yes! Thank you for catching!

**Review by mmeigs — APPROVED** [review]
Thanks!

**Review by seungpark — APPROVED** [review]
LGTM 👍

**Pull request merged**